vyper/builtin_functions/functions.py

```diff
@@ -2156,8 +2158,62 @@ def build_IR(self, expr, context):
         )
 
 
+class ABIDecode(_SimpleBuiltinFunction):
+    _id = "_abi_decode"
+    _inputs = [("data", BytesArrayPrimitive())]
+    _kwargs = {"unwrap_tuple": KwargSettings(BoolDefinition(), True, require_literal=True)}
+
+    def fetch_call_return(self, node):
+        self.infer_arg_types(node)
+
+        if "output_type" not in node._metadata:
+            raise StructureException("Unable to determine the return type of abi_decode", node)
+
+        output_type = node._metadata["output_type"]
+        return output_type
+
+    def infer_arg_types(self, node):
+        self._validate_arg_types(node)
+        data_type = get_exact_type_from_node(node.args[0])
+        return [data_type]
+
+    @validate_inputs
+    def build_IR(self, expr, args, kwargs, context):
+        data = args[0]
+        new_output_typ = self.fetch_call_return(expr)
```

Comment:
there is not really any need for `new_output_type` here - we can just use `output_typ = new_type_to_old_type(self.fetch_call_return(expr))`